File: java/src/main/java/com/pqc/hqc/HqcParams.java
package com.pqc.hqc;

/**
 * Parameter sets for HQC (Hamming Quasi-Cyclic) KEM.
 * <p>
 * HQC is a code-based key encapsulation mechanism selected by NIST for
 * post-quantum cryptography standardization. Each parameter set targets
 * a different NIST security level.
 */
public final class HqcParams {

    /** Parameter set name. */
    public final String name;
    /** Ring dimension (polynomial degree mod x^n - 1). */
    public final int n;
    /** Reed-Solomon codeword length. */
    public final int n1;
    /** Reed-Muller codeword length (duplicated). */
    public final int n2;
    /** Concatenated code length in bits = n1 * n2. */
    public final int n1n2;
    /** Message size in bytes (RS information symbols). */
    public final int k;
    /** RS error correction capability. */
    public final int delta;
    /** RS generator polynomial degree = 2*delta + 1. */
    public final int g;
    /** Weight of secret key vectors x, y. */
    public final int w;
    /** Weight of encryption vectors r1, r2. */
    public final int wr;
    /** Weight of ephemeral error vector e. */
    public final int we;
    /** Public key size in bytes. */
    public final int pkSize;
    /** Secret key size in bytes. */
    public final int skSize;
    /** Ciphertext size in bytes. */
    public final int ctSize;
    /** Shared secret size in bytes. */
    public final int ssSize;

    // Derived sizes
    /** ceil(n / 64). */
    public final int vecNSize64;
    /** ceil(n / 8). */
    public final int vecNSizeBytes;
    /** ceil(n1n2 / 64). */
    public final int vecN1N2Size64;
    /** ceil(n1n2 / 8). */
    public final int vecN1N2SizeBytes;
    /** k bytes. */
    public final int vecKSizeBytes;

    /** GF(2^8) irreducible polynomial. */
    public final int gfPoly;
    /** GF(2^8) multiplicative order = 255. */
    public final int gfMulOrder;

    /** RM(1, rmOrder), base codeword length = 2^rmOrder = 128. */
    public final int rmOrder;
    /** Number of repetitions: n2 / 128. */
    public final int multiplicity;

    // Constants
    /** Seed size used for key generation. */
    public static final int SEED_BYTES = 40;
    /** Size of d = H(m) included in the ciphertext (SHAKE256 output). */
    public static final int HASH_BYTES = 64;
    /** Shared secret size (SHAKE256-512 output). */
    public static final int SHARED_SECRET_BYTES = 64;

    /** Domain separation byte for theta = G(m || pk || salt). */
    public static final byte G_FCT_DOMAIN = 3;
    /** Domain separation byte for d = H(m). */
    public static final byte H_FCT_DOMAIN = 4;
    /** Domain separation byte for ss = K(m || ct). */
    public static final byte K_FCT_DOMAIN = 5;

    private HqcParams(String name, int n, int n1, int n2, int k, int delta,
                      int w, int wr, int we,
                      int pkSize, int skSize, int ctSize,
                      int multiplicity) {
        this.name = name;
        this.n = n;
        this.n1 = n1;
        this.n2 = n2;
        this.n1n2 = n1 * n2;
        this.k = k;
        this.delta = delta;
        this.g = 2 * delta + 1;
        this.w = w;
        this.wr = wr;
        this.we = we;
        this.pkSize = pkSize;
        this.skSize = skSize;
        this.ctSize = ctSize;
        this.ssSize = SHARED_SECRET_BYTES;

        this.vecNSize64 = (n + 63) / 64;
        this.vecNSizeBytes = (n + 7) / 8;
        this.vecN1N2Size64 = (this.n1n2 + 63) / 64;
        this.vecN1N2SizeBytes = (this.n1n2 + 7) / 8;
        this.vecKSizeBytes = k;

        this.gfPoly = 0x11D;
        this.gfMulOrder = 255;
        this.rmOrder = 7;
        this.multiplicity = multiplicity;
    }

    /** HQC-128 targets NIST security level 1 (128-bit). */
    public static final HqcParams HQC_128 = new HqcParams(
        "HQC-128", 17669, 46, 384, 16, 15, 66, 77, 77,
        2249, 2289, 4481, 3
    );

    /** HQC-192 targets NIST security level 3 (192-bit). */
    public static final HqcParams HQC_192 = new HqcParams(
        "HQC-192", 35851, 56, 640, 24, 16, 100, 117, 117,
        4522, 4562, 9026, 5
    );

    /** HQC-256 targets NIST security level 5 (256-bit). */
    public static final HqcParams HQC_256 = new HqcParams(
        "HQC-256", 57637, 90, 640, 32, 29, 131, 153, 153,
        7245, 7285, 14469, 5
    );

    /** Returns all supported HQC parameter sets. */
    public static HqcParams[] allParams() {
        return new HqcParams[]{ HQC_128, HQC_192, HQC_256 };
    }

    @Override
    public String toString() {
        return name;
    }
}

File: dotnet/FnDsa/src/Hqc/HqcParams.cs
namespace FnDsa.Hqc;

/// <summary>
/// Parameter sets for HQC (Hamming Quasi-Cyclic) KEM at three NIST security levels.
/// </summary>
public sealed class HqcParams
{
    public string Name { get; }
    public int N { get; }          // ring dimension (poly degree mod x^n - 1)
    public int N1 { get; }         // Reed-Solomon codeword length
    public int N2 { get; }         // Reed-Muller codeword length (duplicated)
    public int N1N2 { get; }       // concatenated code length in bits = N1 * N2
    public int K { get; }          // message size in bytes (RS information symbols)
    public int Delta { get; }      // RS error correction capability
    public int G { get; }          // RS generator polynomial degree = 2*Delta + 1
    public int W { get; }          // weight of secret key vectors x, y
    public int WR { get; }         // weight of encryption vectors r1, r2
    public int WE { get; }         // weight of ephemeral error vector e
    public int PKSize { get; }     // public key size in bytes
    public int SKSize { get; }     // secret key size in bytes
    public int CTSize { get; }     // ciphertext size in bytes
    public int SSSize { get; }     // shared secret size in bytes

    // Derived sizes
    public int VecNSize64 { get; }       // ceil(N / 64)
    public int VecNSizeBytes { get; }    // ceil(N / 8)
    public int VecN1N2Size64 { get; }
    public int VecN1N2SizeBytes { get; }
    public int VecKSizeBytes { get; }

    // GF(2^8) parameters
    public ushort GFPoly { get; }        // irreducible polynomial for GF(2^8)
    public int GFMulOrder { get; }       // multiplicative order = 255

    // Reed-Muller parameters
    public int RMOrder { get; }          // RM(1, RMOrder), base codeword length = 2^RMOrder = 128
    public int Multiplicity { get; }     // number of repetitions: N2 / 128

    private HqcParams(
        string name, int n, int n1, int n2, int k, int delta,
        int w, int wr, int we,
        int pkSize, int skSize, int ctSize,
        int multiplicity)
    {
        Name = name;
        N = n;
        N1 = n1;
        N2 = n2;
        N1N2 = n1 * n2;
        K = k;
        Delta = delta;
        G = 2 * delta + 1;
        W = w;
        WR = wr;
        WE = we;
        PKSize = pkSize;
        SKSize = skSize;
        CTSize = ctSize;
        SSSize = SharedSecretBytes;

        VecNSize64 = (n + 63) / 64;
        VecNSizeBytes = (n + 7) / 8;
        VecN1N2Size64 = (N1N2 + 63) / 64;
        VecN1N2SizeBytes = (N1N2 + 7) / 8;
        VecKSizeBytes = k;

        GFPoly = 0x11D;
        GFMulOrder = 255;
        RMOrder = 7;
        Multiplicity = multiplicity;
    }

    // Constants
    public const int SeedBytes = 40;
    public const int HashBytes = 64;
    public const int SharedSecretBytes = 64;

    // Domain separation bytes for SHAKE-256 hashing
    public const byte GFctDomain = 3;  // theta = G(m || pk || salt)
    public const byte HFctDomain = 4;  // d = H(m)
    public const byte KFctDomain = 5;  // ss = K(m || ct)

    /// <summary>HQC-128: NIST security level 1 (128-bit).</summary>
    public static readonly HqcParams HQC128 = new(
        name: "HQC-128", n: 17669, n1: 46, n2: 384, k: 16, delta: 15,
        w: 66, wr: 77, we: 77,
        pkSize: 2249, skSize: 2289, ctSize: 4481,
        multiplicity: 3);

    /// <summary>HQC-192: NIST security level 3 (192-bit).</summary>
    public static readonly HqcParams HQC192 = new(
        name: "HQC-192", n: 35851, n1: 56, n2: 640, k: 24, delta: 16,
        w: 100, wr: 117, we: 117,
        pkSize: 4522, skSize: 4562, ctSize: 9026,
        multiplicity: 5);

    /// <summary>HQC-256: NIST security level 5 (256-bit).</summary>
    public static readonly HqcParams HQC256 = new(
        name: "HQC-256", n: 57637, n1: 90, n2: 640, k: 32, delta: 29,
        w: 131, wr: 153, we: 153,
        pkSize: 7245, skSize: 7285, ctSize: 14469,
        multiplicity: 5);

    /// <summary>Returns all supported HQC parameter sets.</summary>
    public static HqcParams[] All() => [HQC128, HQC192, HQC256];
}
